google/resource_compute_instance.go

```diff
@@ -1032,9 +1058,10 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 				"name":               iface.Name,
 				"address":            iface.NetworkIP,
 				"network":            iface.Network,
-				"subnetwork":         iface.Subnetwork,
+				"subnetwork":         ConvertSelfLinkToV1(iface.Subnetwork),
```

Comment:
Is the `DiffSuppressFunc` not covering this one?

Reply:
The problem is that subnetwork can be specified using the `name` or the `self-link`.

If a user specify a name in the tf config, the diff is suppressed correctly using `linkDiffSuppress`.
However, if the user specify a self-link in the tf config, we get a diff because one link is beta and the other one is v1 and linkDiffSuppress doesn't work for that case.

The other option is to change the DiffSuppressFunc to something along these lines:
 ```go
DiffSuppressFunc: func(...) bool { 
    return compareSelfLinkRelativePaths(...) || linkDiffSuppress(...)
}
```

We could factored this logic into its own method `compareRegionalSelfLinkOrResourceName` or something like that. The same way it was done for `compareGlobalSelfLinkOrResourceName`

Reply:
That makes sense, thanks!

Looking at how `linkDiffSuppress` is used for both scopes, would we be able to generalise `compareGlobalSelfLinkOrResourceName`, and use that instead?

As I understand it, we are always going to store a full self link in state, so `old` will always be a full link; we can then split `new` on `/`, and check its length, where there are two cases:

- `newParts` has a length of 1: The user has specified a resource `name`, so we should split `old` into `oldParts` and check the last element of each.
- `newParts` has any other length: The user has not specified a name, and so they are using a relative self link (starting from `projects/`), or a full self link; we can compare by using `compareSelfLinkRelativePaths` on `old` and `new`.

That way we can cover both cases with the same diff function, and this should work for Zonal resources too if there is a similar case, without adding a third function for that as well.

Reply:
Done